src/main/daml/Daml/Finance/Interface/Util/SyntheticKey.daml

```diff
@@ -0,0 +1,254 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+module Daml.Finance.Interface.Util.SyntheticKey where
+
+import DA.Assert ((===))
+import DA.Map qualified as M (Map, delete, empty, insert, lookup)
+import DA.Optional (fromSomeNote)
+import DA.Text (sha256)
+import DA.Traversable qualified as T (mapA)
+import Daml.Control.Monad.Trans.Class
+import Daml.Control.Monad.Trans.State
+import Prelude hiding (archive, create, exercise, exerciseByKey, fetch, fetchByKey, fetchFromInterface, lookup, lookupByKey)
+import qualified Prelude
+
+-- | Type synonym for `SyntheticKey`.
+type I = SyntheticKey
+
+-- | Type synonym for `View`.
+type V = View
+
+-- | An interface for synthetically keyed interfaces.
+interface SyntheticKey where
+  viewtype V
+
+-- | View for `I`.
+data View = View {}
+  deriving (Eq, Show)
+
+-- ============= State: Generic Index Table =============
+-- | Index table (used for looking up contract ids for keys).
+newtype IndexTable = IndexTable {getIndexTable : M.Map Text (ContractId SyntheticKey)} deriving (Eq,Show)
+
+-- | HIDE
+class
+  ( Eq k
+  , Show k
+  , Ord k
+  , HasFetch i
+  , HasInterfaceTypeRep i
+  , HasFromInterface i SyntheticKey
+  , HasToInterface i I
+  ) =>
+  HasSyntheticKey i k where
+    getKey : i -> k
+    getKeyFromCid : Party -> ContractId i -> Update k
```

Comment:
Can we document what these two methods are supposed to do? 

In the case of `getKeyFromCid` we should also specify how this is usually authorized (by exercising a choice with flexible controller)